[Bug] Gemma4 VLM from_pretrained(torch_dtype=float32) leaves nested submodules in bf16 and breaks FSDP2

## Summary

`NeMoAutoModelForImageTextToText.from_pretrained(..., torch_dtype=torch.float32)` does not build `google/gemma-4-E4B-it` with uniform `float32` parameters.

The top-level HF config is changed to `float32`, but nested Gemma4 configs (`text_config`, `vision_config`, `audio_config`) keep the checkpoint dtype (`bfloat16`). As a result, modules built through `AutoModel.from_config(sub_config)` become `bfloat16`, while directly constructed modules such as `embed_vision`, `embed_audio`, and `lm_head` become `float32`.

FSDP2 then fails on the first forward:

```text
AssertionError: FSDP expects uniform original parameter dtype but got {torch.bfloat16, torch.float32}
```

## Reproduction

Start from `examples/vlm_finetune/gemma4/gemma4_4b_mock.yaml`:

```diff
-  torch_dtype: torch.bfloat16
+  torch_dtype: torch.float32
   ...
 freeze_config:
-  freeze_vision_tower: true
+  freeze_vision_tower: false
```

Run:

```bash
torchrun --nproc-per-node=8 --nnodes=1 \
  nemo_automodel/recipes/vlm/finetune.py \
  -c examples/vlm_finetune/gemma4/gemma4_4b_mock_vlm_repro.yaml
```

Observed:

```text
File "torch/distributed/fsdp/_fully_shard/_fsdp_param_group.py", line 238, in _init_mp_dtypes
AssertionError: FSDP expects uniform original parameter dtype but got {torch.bfloat16, torch.float32}
```

## Root Cause

`nemo_automodel/_transformers/model_init.py` currently applies the requested dtype only to the top-level config:

```python
if torch_dtype != "auto":
    hf_config.torch_dtype = torch_dtype
```

For Gemma4 VLM this misses nested configs used by HF during construction, for example [google/gemma-4-E4B-it](https://huggingface.co/google/gemma-4-E4B-it):

- `language_model`, `vision_tower`, and `audio_tower` use `AutoModel.from_config(sub_config)` and read the nested `torch_dtype`.
- `embed_vision`, `embed_audio`, and `lm_head` are direct `nn.Linear` modules and inherit `torch.get_default_dtype()`.

Temporary dtype probes show the mismatch exists immediately after construction:

```text
requested torch_dtype:      torch.float32
top.torch_dtype:           torch.float32
text_config.torch_dtype:   torch.bfloat16
vision_config.torch_dtype: torch.bfloat16
audio_config.torch_dtype:  torch.bfloat16

after_construct: [('torch.bfloat16', 1158), ('torch.float32', 3)]
float32 params:
  model.embed_vision.embedding_projection.weight
  model.embed_audio.embedding_projection.weight
  lm_head.weight
```

The checkpoint itself is uniformly `BF16`; the mixed dtype is introduced during model construction.

## Expected Behavior

`from_pretrained(torch_dtype=X)` should construct all parameters with dtype `X`, including parameters controlled by nested multimodal sub-configs. With `torch_dtype=torch.float32`, FSDP2's uniform original-dtype check should pass.

## Possible Fix

One possible fix is to propagate explicit dtype overrides into nested multimodal configs before constructing the model:

```python
if torch_dtype != "auto":
    hf_config.torch_dtype = torch_dtype

    for sub_config_name in ("text_config", "vision_config", "audio_config"):
        sub_config = getattr(hf_config, sub_config_name, None)
        if sub_config is not None and hasattr(sub_config, "torch_dtype"):
            sub_config.torch_dtype = torch_dtype
```

With this local fix, the same reproducer constructs uniformly:

```text
text_config.torch_dtype:   torch.float32
vision_config.torch_dtype: torch.float32
audio_config.torch_dtype:  torch.float32
```

--- a/nemo_automodel/_transformers/model_init.py
+++ b/nemo_automodel/_transformers/model_init.py
@@ -148,6 +148,36 @@ def local_torch_dtype(
         torch.set_default_dtype(original_dtype)
 
 
+def _propagate_torch_dtype_to_subconfigs(hf_config, torch_dtype: torch.dtype) -> None:
+    """Recursively set ``torch_dtype`` on ``hf_config`` and all nested sub-configs.
+
+    Multimodal configs (e.g. ``Gemma4ForConditionalGeneration``) hold nested
+    sub-configs such as ``text_config``, ``vision_config``, and ``audio_config``.
+    During model construction, HF builds sub-modules via
+    ``AutoModel.from_config(sub_config)``, which reads ``torch_dtype`` from the
+    sub-config rather than the parent. Without propagation, those sub-modules
+    keep the checkpoint dtype while directly instantiated ``nn.Linear`` modules
+    take the requested default dtype, producing a mixed-dtype model that FSDP2's
+    uniform-original-dtype check rejects.
+
+    Args:
+        hf_config: The top-level HuggingFace config to update in place.
+        torch_dtype: The dtype to assign to every nested ``PretrainedConfig``.
+    """
+    seen: set[int] = set()
+
+    def _recurse(cfg) -> None:
+        if id(cfg) in seen:
+            return
+        seen.add(id(cfg))
+        cfg.torch_dtype = torch_dtype
+        for value in vars(cfg).values():
+            if isinstance(value, PretrainedConfig):
+                _recurse(value)
+
+    _recurse(hf_config)
+
+
 def _is_config_compatible_with_custom_model(arch_name: str, config) -> bool:
     """
     Check if a HuggingFace config is compatible with our custom model implementation.
@@ -650,6 +680,15 @@ def __init_model(
     )
     architectures = get_architectures(hf_config)
 
+    # Propagate the user-requested dtype to the top-level config and every nested
+    # sub-config (text/vision/audio). Multimodal models like Gemma4 build their
+    # sub-towers via AutoModel.from_config(sub_config), which reads torch_dtype
+    # from the sub-config; without this, sub-towers stay at the checkpoint dtype
+    # while directly instantiated modules (lm_head, embed_vision, embed_audio)
+    # take the requested dtype, tripping FSDP2's uniform-dtype check.
+    if torch_dtype != "auto":
+        _propagate_torch_dtype_to_subconfigs(hf_config, torch_dtype)
+
     # Streaming BnB loading: when quantization is requested and we're loading from a
     # pretrained checkpoint, use streaming quantization to avoid materializing the full
     # bf16 model in memory. This is critical for unified-memory systems (DGX Spark)
@@ -740,9 +779,6 @@ def __init_model(
                 from nemo_automodel.components.models.common.utils import BackendConfig
 
                 kwargs["backend"] = BackendConfig(**kwargs["backend"])
-            # Override config's torch_dtype with user-requested dtype so model __init__ uses correct dtype
-            if torch_dtype != "auto":
-                hf_config.torch_dtype = torch_dtype
             with local_torch_dtype(torch_dtype, model_cls.__name__):
                 return True, model_cls(hf_config, *model_args, **kwargs)
 

--- a/tests/unit_tests/_transformers/test_model_init.py
+++ b/tests/unit_tests/_transformers/test_model_init.py
@@ -26,6 +26,7 @@ from nemo_automodel._transformers.model_init import (
     _has_safetensors,
     _init_model,
     _load_config_with_layer_types_fix,
+    _propagate_torch_dtype_to_subconfigs,
     _resolve_model_dir,
     _setup_bnb_loading_kwargs,
     _stream_load_bnb_weights,
@@ -72,6 +73,61 @@ class TestConsumeConfigOverridesNestedDict:
 
         assert config.some_field == {"key": "val"}
         assert "some_field" not in kwargs
+
+
+class TestPropagateTorchDtypeToSubconfigs:
+    """Nested PretrainedConfig sub-configs must receive the requested torch_dtype."""
+
+    def test_propagates_to_nested_multimodal_subconfigs(self):
+        """Gemma4-style VLM configs expose text/vision/audio sub-configs that all need updating."""
+        from transformers import PretrainedConfig
+
+        text_config = PretrainedConfig()
+        text_config.torch_dtype = torch.bfloat16
+        vision_config = PretrainedConfig()
+        vision_config.torch_dtype = torch.bfloat16
+        audio_config = PretrainedConfig()
+        audio_config.torch_dtype = torch.bfloat16
+
+        top = PretrainedConfig()
+        top.torch_dtype = torch.bfloat16
+        top.text_config = text_config
+        top.vision_config = vision_config
+        top.audio_config = audio_config
+
+        _propagate_torch_dtype_to_subconfigs(top, torch.float32)
+
+        assert top.torch_dtype == torch.float32
+        assert text_config.torch_dtype == torch.float32
+        assert vision_config.torch_dtype == torch.float32
+        assert audio_config.torch_dtype == torch.float32
+
+    def test_ignores_non_config_attributes(self):
+        """Non-config attributes (e.g. plain dicts, ints) must not be traversed or mutated."""
+        from transformers import PretrainedConfig
+
+        top = PretrainedConfig()
+        top.torch_dtype = torch.bfloat16
+        top.some_dict = {"not": "a_config"}
+        top.hidden_size = 4096
+
+        _propagate_torch_dtype_to_subconfigs(top, torch.float32)
+
+        assert top.torch_dtype == torch.float32
+        assert top.some_dict == {"not": "a_config"}
+        assert top.hidden_size == 4096
+
+    def test_handles_cycles(self):
+        """Self-referencing configs must not cause infinite recursion."""
+        from transformers import PretrainedConfig
+
+        top = PretrainedConfig()
+        top.torch_dtype = torch.bfloat16
+        top.self_ref = top
+
+        _propagate_torch_dtype_to_subconfigs(top, torch.float32)
+
+        assert top.torch_dtype == torch.float32
 
 
 class TestBackendDictCoercion:

